contracts/golos.vesting/golos.vesting.cpp

```diff
@@ -0,0 +1,441 @@
+            item.percentage_deductions = percentage_deductions;
+            item.return_date = time_point_sec(now() + OUTPUT_PAYOUT_PERIOD);
+//            item.return_date = time_point_sec(now()); // TODO test
+        });
+    }
+
+    tables::account_table account_recipient(_self, recipient);
+    auto balance_recipient = account_recipient.find(quantity.symbol.name());
+    if (balance_recipient != account_recipient.end()) {
+        account_recipient.modify(balance_recipient, sender, [&](auto &item){
+            item.delegate_vesting += quantity;
+        });
+    } else {
+        account_recipient.emplace(sender, [&](auto &item){
+            item.delegate_vesting = quantity;
+            item.vesting.symbol = quantity.symbol;
+            item.received_vesting.symbol = quantity.symbol;
+        });
+    }
+}
+
+void vesting::undelegate_vesting(account_name sender, account_name recipient, asset quantity) {
+    require_auth(sender);
+
+    tables::delegate_table table(_self, quantity.symbol.name());
+    auto index_table = table.get_index<N(sender)>();
+    auto it_index = index_table.find(sender);
+    eosio_assert(it_index != index_table.end(), "Not enough delegated vesting");
+
+    while(it_index != index_table.end() && it_index->sender == sender) {
+        if(it_index->recipient == recipient) {
+            eosio_assert(it_index->return_date <= time_point_sec(now()), "Tokens are frozen until the end of the period");
+            eosio_assert(it_index->quantity >= quantity, "There are not enough delegated tools for output");
+
+            asset part_deductions;
+            if (it_index->quantity == quantity) {
+                part_deductions = it_index->deductions;
+                index_table.erase(it_index);
+            } else {
+                 auto persent = (quantity * FULL_PERSENT) / it_index->quantity;
+                 part_deductions = it_index->deductions * persent / FULL_PERSENT;
+
+                index_table.modify(it_index, sender, [&](auto &item){
+                    item.quantity -= quantity;
+                    item.deductions -= part_deductions;
+                });
+            }
+
+            _table_delegate_vesting.emplace(sender, [&](structures::return_delegate &item){
+                item.id = _table_delegate_vesting.available_primary_key();
+                item.recipient = sender;
+                item.amount = quantity + part_deductions;
+                item.date = time_point_sec(now() + OUTPUT_PAYOUT_PERIOD);
+            });
+
+            tables::account_table account_recipient(_self, recipient);
+            auto balance = account_recipient.find(quantity.symbol.name());
+            eosio_assert(balance != account_recipient.end(), "This token is not on the balance sheet");
+            account_recipient.modify(balance, sender, [&](auto &item){
+                item.delegate_vesting -= quantity;
+            });
+
+            break;
+        }
+
+        ++it_index;
+    }
+}
```

Comment:
I think this code needs return error if sender doesn't delegate vesting to recipient.